controllers/src/ComputedTorqueFixedBase.cpp

```diff
@@ -0,0 +1,271 @@
+#include <iDynTree/Core/VectorDynSize.h>
+#include <iDynTree/Core/VectorFixSize.h>
+#include <iDynTree/KinDynComputations.h>
+#include <iDynTree/Model/FreeFloatingState.h>
+#include <iDynTree/ModelIO/ModelLoader.h>
+
+using namespace gympp::controllers;
+
+class ComputedTorqueFixedBase::Impl
+{
+public:
+    class Buffers;
+
+    std::string urdfFile;
+    gympp::RobotPtr robot;
+    std::vector<std::string> controlledJoints;
+
+    std::vector<double> kpInitial;
+    std::vector<double> kdInitial;
+    std::unordered_map<std::string, JointControlMode> initialControlMode;
+
+    PositionControllerReferences references;
+    std::unique_ptr<Buffers> buffers;
+    std::unique_ptr<iDynTree::KinDynComputations> kinDyn;
+
+    static Eigen::Map<Eigen::VectorXd> toEigen(const std::vector<double>& vector)
+    {
+        return {const_cast<double*>(vector.data()), Eigen::Index(vector.size())};
+    }
+};
+
+class ComputedTorqueFixedBase::Impl::Buffers
+{
+public:
+    Buffers(const unsigned controlledDofs = 0)
+    {
+        jointPositions.resize(controlledDofs);
+        jointVelocities.resize(controlledDofs);
+        massMatrix.resize(controlledDofs + 6, controlledDofs + 6);
+
+        kp = Eigen::ArrayXd(controlledDofs);
+        kd = Eigen::ArrayXd(controlledDofs);
+
+        torques = Eigen::VectorXd(controlledDofs);
+        s_ddot_star = Eigen::VectorXd(controlledDofs);
+        positionError = Eigen::VectorXd(controlledDofs);
+        velocityError = Eigen::VectorXd(controlledDofs);
+    }
+
+    iDynTree::Vector3 gravity;
+    iDynTree::MatrixDynSize massMatrix;
+    iDynTree::VectorDynSize jointPositions;
+    iDynTree::VectorDynSize jointVelocities;
+    iDynTree::FreeFloatingGeneralizedTorques biasForces;
+
+    Eigen::ArrayXd kp;
+    Eigen::ArrayXd kd;
+
+    Eigen::VectorXd torques;
+    Eigen::VectorXd s_ddot_star;
+    Eigen::VectorXd positionError;
+    Eigen::VectorXd velocityError;
+};
+
+ComputedTorqueFixedBase::ComputedTorqueFixedBase(const std::string& urdfFile,
+                                                 gympp::RobotPtr robot,
+                                                 const std::vector<double>& kp,
+                                                 const std::vector<double>& kd,
+                                                 const std::vector<std::string>& controlledJoints)
+    : pImpl{std::make_unique<Impl>()}
+{
+    pImpl->robot = robot;
+    pImpl->urdfFile = urdfFile;
+    pImpl->controlledJoints = controlledJoints;
+    assert(robot->valid());
+    assert(!pImpl->urdfFile.empty());
+    assert(robot->dofs() == controlledJoints.size() || controlledJoints.empty());
+
+    pImpl->kpInitial = kp;
+    pImpl->kdInitial = kd;
+    assert(robot->dofs() == kp.size());
+    assert(kp.size() == kd.size());
+}
+
+ComputedTorqueFixedBase::~ComputedTorqueFixedBase() = default;
+
+bool ComputedTorqueFixedBase::initialize()
+{
+    gymppDebug << "Initializing ComputedTorqueFixedBaseCpp" << std::endl;
```

Comment:
Leftover?

Reply:
Ah, I just realized it is a draft, so it make sense to just have the messages for now.